docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,260 @@
+
+In the SDK every time a `Coin` is initialized through the constructor function `NewCoin`, a validation
+of a coin's denom is performed according to a
+[Regex](https://github.com/cosmos/cosmos-sdk/blob/a940214a4923a3bf9a9161cd14bd3072299cd0c9/types/coin.go#L583),
+where only lowercase alphanumeric characters are accepted. While this is desirable for native denoms
+to keep a clean UX, it presents a challenge for IBC as ports and channels might be randomly
+generated with special carracters and uppercases as per the [ICS 024 - Host
+Requirements](https://github.com/cosmos/ics/tree/master/spec/ics-024-host-requirements#paths-identifiers-separators)
+specification.
+
+## Decision
+
+Instead of adding the identifiers on the coin denomination directly, the proposed solution hashes
+the denomination prefix in order to get a consistent length for all the cross-chain fungible tokens.
+The new format will be the following:
+
+```golang
+ibcDenom = "ibc/" + SHA256 hash of the trace identifiers prefix + "/" + base coin denomination
+```
+
+### `x/ibc-transfer` Changes
+
+In order to retreive the trace information from an IBC denomination, a lookup table needs to be
+added to the `ibc-transfer` module. These values need to also be persisted between upgrades, meaning
+that a new `[]Trace` `GenesisState` field state needs to be added to the module:
+
+```golang
+// GetDenom retreives the full identifiers trace from the store.
+func (k Keeper) GetTrace(ctx Context, traceHash []byte) string {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyTrace(traceHash))
+  if len(bz) == 0 {
+    return ""
+  }
+  return string(bz)
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasTrace(ctx Context, traceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(traceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, traceHash []byte, trace string) {
+  store := ctx.KVStore(k.storeKey)
+  store.Set(types.KeyTrace(traceHash), []byte(trace))
+}
+```
+
+When a fungible token is send to a sink chain, the trace information needs to be updated with the
+new port and channel identifiers:
+
+```golang
+// PrefixDenom adds the given port and channel identifiers prefix to the denomination and sets the
+// new {trace hash -> trace} pair to the store.
+func (k Keeper) PrefixDenom(ctx Context, portID, channelID, denom string) string {
+  // Get each component of the denom. The resulting slice will be:
+  //
+  // - ["ibc", traceHash, baseDenom], if the denom is dirty (contains trace metadata).
+  // - [baseDenom], if the denom has never been sent from the origin chain.
+  denomSplit := strings.SplitN(denom, "/", 3)
+
+  var (
+    baseDenom string
+    trace     string
+    traceHash tmbytes.HexBytes
+  )
+
+  // check if the denomination is clean or if it contains the trace info
+  if denomSplit[0] == denom {
```

Comment:
it will always be according to the `Split` godoc. Do you want me to add it as a safety check?

Reply:
No because I want to ensure that it is only one element and not more than that.

Otherwise a denomSplit `atom/notatom` will enter this conditional if the passed-in denom is `atom`